File: app/src/main/java/iuliiaponomareva/evroscudo/CurrencyAdapter.java
package iuliiaponomareva.evroscudo;

import android.content.Context;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import androidx.annotation.NonNull;


public class CurrencyAdapter extends ArrayAdapter<Currency> {
    private DisplayRatesActivity activity;

    public CurrencyAdapter(DisplayRatesActivity activity, int resource) {
        super(activity, resource);
        this.activity = activity;
    }

    static class ViewHolder {
        TextView codeTextView;
        TextView bankRateTextView;
        TextView bankRateTextView2;
    }

    @Override
    @NonNull
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        ViewHolder holder;
        if (convertView == null) {
            LayoutInflater inflater = (LayoutInflater)activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = inflater.inflate(R.layout.list_item, parent, false);
            holder = new ViewHolder();
            holder.codeTextView = convertView.findViewById(R.id.currency_code);
            holder.bankRateTextView = convertView.findViewById(R.id.bank_rate);
            holder.bankRateTextView2 = convertView.findViewById(R.id.bank_2_rate);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }
        String code = getItem(position).getCode();
        holder.codeTextView.setText(code);
        Bank bank1 = activity.getFirstBank();
        String rate1 = getItem(position).getBankRate(bank1.getBanks());
        showRate(position, holder.bankRateTextView, code, bank1, rate1);
        Bank bank2 = activity.getSecondBank();
        String rate2 = getItem(position).getBankRate(bank2.getBanks());
        showRate(position, holder.bankRateTextView2, code, bank2, rate2);

        return convertView;
    }

    private void showRate(int position, TextView bankRateTextView, String code, Bank bank, String rate) {
        if (TextUtils.isEmpty(rate))
            bankRateTextView.setText("");
        else {
            StringBuilder text = new StringBuilder(rate);
            text.append(" ");
            if (bank.inMyCurrency) {
                text.append(bank.getCurrencyCode());
                text.append(activity.getString(R.string.is));
                text.append(getItem(position).getNominal(bank.getBanks()));
            } else {
                text.append(code);
                text.append(activity.getString(R.string.is));
                text.append(getItem(position).getNominal(bank.getBanks()));
                text.append(bank.getCurrencyCode());
            }
            bankRateTextView.setText(text);
        }
    }
}

File: app/src/main/java/iuliiaponomareva/evroscudo/CurrencyAdapter.kt
package iuliiaponomareva.evroscudo

import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.RecyclerView
import java.util.*


class CurrencyAdapter(private val activity: DisplayRatesActivity) :
    RecyclerView.Adapter<CurrencyAdapter.CurrencyViewHolder>() {
    private val rates = ArrayList<Currency>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CurrencyViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.list_item, parent, false)
        return CurrencyViewHolder(view)
    }

    override fun onBindViewHolder(holder: CurrencyViewHolder, position: Int) {
        val code = rates[position].code
        holder.codeTextView.text = code
        val bank1 = activity.firstBank
        val rate1 = rates[position].getBankRate(bank1.bankId)
        showRate(position, holder.bankRateTextView, code, bank1, rate1)
        val bank2 = activity.secondBank
        val rate2 = rates[position].getBankRate(bank2.bankId)
        showRate(position, holder.bankRateTextView2, code, bank2, rate2)
    }

    override fun getItemCount(): Int {
        return rates.size
    }

    fun set(currencies: Collection<Currency>) {
        val needed = currencies
            .filter {
                PreferenceManager.getDefaultSharedPreferences(this.activity)
                    .getBoolean(it.code, true)
            }
            .filter {
                (it.getBankRate(activity.firstBank.bankId) != null) || (it.getBankRate(
                    activity.secondBank.bankId
                ) != null)
            }
        rates.clear()
        rates.addAll(needed)
        notifyDataSetChanged()
    }

    private fun showRate(
        position: Int,
        bankRateTextView: TextView,
        code: String,
        bank: Bank,
        rate: String?
    ) {
        if (TextUtils.isEmpty(rate))
            bankRateTextView.text = ""
        else {
            val text = StringBuilder(rate!!)
            text.append(" ")
            if (bank.inMyCurrency) {
                text.append(bank.currencyCode)
                text.append(activity.getString(R.string.`is`))
                text.append(rates[position].getNominal(bank.bankId))
            } else {
                text.append(code)
                text.append(activity.getString(R.string.`is`))
                text.append(rates[position].getNominal(bank.bankId))
                text.append(bank.currencyCode)
            }
            bankRateTextView.text = text
        }
    }

    class CurrencyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var codeTextView: TextView
        var bankRateTextView: TextView
        var bankRateTextView2: TextView

        init {
            codeTextView = itemView.findViewById(R.id.currency_code)
            bankRateTextView = itemView.findViewById(R.id.bank_rate)
            bankRateTextView2 = itemView.findViewById(R.id.bank_2_rate)
        }
    }
}
